### PR TITLE
feat: support custom dialer

### PR DESCRIPTION
**Pull request by mar4uk**
This is basically a copy of https://github.com/bradfitz/gomemcache/pull/86 with a fixing review comment

### CODE REVIEW

**Comment by mar4uk** [comment]
@bradfitz if everything is ok, could you please merge this one?

File: memcache/memcache.go [hunk 1]
```diff
@@ -132,6 +132,9 @@ func NewFromSelector(ss ServerSelector) *Client {
 // Client is a memcache client.
 // It is safe for unlocked use by multiple concurrent goroutines.
 type Client struct {
+	// Dialer specifies a custom dialer used to dial new connections to a server.
```
Comment (bradfitz) [inline memcache/memcache.go:135]: `// DialTimeout`, not `// Dialer`.

But the more idiomatic thing to do these days would be to use `DialContext func(ctx context.Context, network, address string) (Conn, error)` and have us pass in a context instead.

**Review by bradfitz — CHANGES_REQUESTED** [review]
(Let's use DialContext instead)

**Comment by mar4uk** [comment]
@bradfitz Thank you for your review! It definitely makes sense to use DialContext. I did changes, please take a look when have time

**Review by bradfitz — APPROVED** [review]

**Pull request merged**